specs/api_cron_flow.md

```diff
@@ -0,0 +1,82 @@
+# Automating the API Update Flow with Cron
+
+## Problem Statement
+
+Some of the extern APIs that we use may not allow us to process all of
+our data records in a single batch. If we send too many requests to
+the API in a short time, our access may be throttled or suspended.
+
+## Background and Significance
+
+
+There are currently two APIs that we use to enhance the data sets from
+the SBA.
+
+The first is an API to get the Yelp rating where possible.
+
+The second is a Google Civics API to get the Congressional district
+based on the borrower address.
+
+Given the size of our database (even just for the SFDO loan set), if
+we try to update all the ratings in a single run, the API providers
+may throttle our access.
+
+The request is to be able to chunk the output into small batches and
+schedule a cron job to run repeatedly until all records have been
+attempted.
+
+## Final Result
+
+For this project, we will make changes to the script to allow it to
+run a specific set of records and then exit. We will provide the
+capability to keep track of which records have been attempted and
+which have not. We will also provide an example of a cron scheduling
+job that would permit the entire database to be refreshed over a
+period of hours or days.
+
+## Proposed Approach
+
+### Updates to the API control script
+
+The API control script will need a new argument which indicates the
+maximum number of calls it should make to each API it supports. This
+could be named something like '--max_attempts'. I suggest another
+argument to force it to restart with the first record, something like
+'--reset_last_attempt'.
```

Comment:
+1 to force reset, that's a good idea.  I'm also assuming there'll be an argument like --api=yelp or --api=geocoder indicating which API the script should hit, so that the crontab can have more granular control, e.g. fetch Yelp ratings every day, but fetch congressional districts once a month, or something like that.

Reply:
Hadn't thought about a switch to run just one API at a time, but it make sense and it easy to do.